BridgeAppSDK/SBAJSONObject.m

```diff
@@ -0,0 +1,257 @@
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#import "SBAJSONObject.h"
+
+
+/**
+ Sage requires our dates to be in "ISO 8601" format,
+ like this:
+ 
+ 2015-02-25T16:42:11+00:00
+ 
+ Got the rules from http://en.wikipedia.org/wiki/ISO_8601
+ Date-formatting rules from http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns
+ */
+static NSString * const kDateFormatISO8601 = @"yyyy-MM-dd'T'HH:mm:ssZZZZZ";
+
+@implementation NSString (SBAJSONObject)
+
+- (id)jsonObjectWithFormatter:(NSFormatter * _Nullable) formatter  {
+    if ([formatter isKindOfClass:[NSNumberFormatter class]]) {
+        return [(NSNumberFormatter *)formatter numberFromString:self];
+    }
+    else {
+        return [self copy];
+    }
+}
+
+- (NSNumber *)boolNumber {
+    if ([self compare: @"no" options: NSCaseInsensitiveSearch] == NSOrderedSame ||
+        [self compare: @"false" options: NSCaseInsensitiveSearch] == NSOrderedSame)
+    {
+        return @(NO);
+    }
+    
+    else if ([self compare: @"yes" options: NSCaseInsensitiveSearch] == NSOrderedSame ||
+             [self compare: @"true" options: NSCaseInsensitiveSearch] == NSOrderedSame)
+    {
+        return @(YES);
+    }
+    
+    return nil;
+}
+
+- (NSNumber *)intNumber {
+    NSInteger itemAsInt = [self integerValue];
+    NSString *verificationString = [NSString stringWithFormat: @"%d", (int) itemAsInt];
+    
+    // Here, we use -isValidJSONObject: to make sure the int isn't
+    // NaN or infinity.  According to the JSON rules, those will
+    // break the serializer.
+    if ([verificationString isEqualToString: self] && [NSJSONSerialization isValidJSONObject: @[verificationString]])
+    {
+        return @(itemAsInt);
+    }
+    
+    return nil;
+}
+
+@end
+
+@implementation NSNumber (SBAJSONObject)
+
+- (id)jsonObjectWithFormatter:(NSFormatter * _Nullable) formatter  {
+    if ([formatter isKindOfClass:[NSNumberFormatter class]]) {
+        return [(NSNumberFormatter *)formatter stringFromNumber:self];
+    }
+    else {
+        return [self copy];
+    }
+}
+
+@end
+
+@implementation NSNull (SBAJSONObject)
+
+- (id)jsonObjectWithFormatter:(NSFormatter * _Nullable) __unused formatter  {
+    return self;
+}
+
+@end
+
+@implementation NSDate (SBAJSONObject)
+
+- (id)jsonObjectWithFormatter:(NSFormatter * _Nullable)formatter {
+    NSDateFormatter *dateFormatter = [formatter isKindOfClass:[NSDateFormatter class]] ? (NSDateFormatter *)formatter : [self defaultFormatter];
+    return [dateFormatter stringFromDate:self];
+}
+
+- (NSDateFormatter *)defaultFormatter {
```

Comment:
BridgeSDK has a public category (NSDate+SBBAdditions.h) to handle Bridge-compliant conversions between NSDate objects and ISO8601 date strings. Possible to splice in here instead of duplicating that functionality?